### DISCUSSION TITLE
model训练完后，测试model。执行det没问题，rec报错

### DISCUSSION THREAD

**Discussion by Advance-Mr** [PaddlePaddle/PaddleOCR#14327] [Q&A]
### 🔎 Search before asking

- [X] I have searched the PaddleOCR [Docs](https://paddlepaddle.github.io/PaddleOCR/) and found no similar bug report.
- [X] I have searched the PaddleOCR [Issues](https://github.com/PaddlePaddle/PaddleOCR/issues) and found no similar bug report.
- [X] I have searched the PaddleOCR [Discussions](https://github.com/PaddlePaddle/PaddleOCR/discussions) and found no similar bug report.

### 🐛 Bug (问题描述)

(paddle2) E:\PaddleOCR-release-2.6>python tools/infer/predict_system.py --image_dir="C:\Users\Administrator\Desktop\6.jpg" --det_model_dir="./inference_model/det/" --rec_model_dir="./inference_model/rec/"
[2024/11/25 12:21:22] ppocr INFO: In PP-OCRv3, rec_image_shape parameter defaults to '3, 48, 320', if you are using recognition model with PP-OCRv2 or an older version, please set --rec_image_shape='3,32,320
[2024/11/25 12:21:24] ppocr DEBUG: dt_boxes num : 11, elapsed : 1.905656337738037
Traceback (most recent call last):
  File "tools/infer/predict_system.py", line 272, in <module>
    main(args)
  File "tools/infer/predict_system.py", line 190, in main
    dt_boxes, rec_res, time_dict = text_sys(img)
  File "tools/infer/predict_system.py", line 105, in __call__
    rec_res, elapse = self.text_recognizer(img_crop_list)
  File "E:\PaddleOCR-release-2.6\tools\infer\predict_rec.py", line 623, in __call__
    rec_result = self.postprocess_op(preds, return_word_box=self.return_word_box, wh_ratio_list=wh_ratio_list, max_wh_ratio=max_wh_ratio)
  File "E:\PaddleOCR-release-2.6\ppocr\postprocess\rec_postprocess.py", line 170, in __call__
    text = self.decode(preds_idx, preds_prob, is_remove_duplicate=True, return_word_box=return_word_box)
  File "E:\PaddleOCR-release-2.6\ppocr\postprocess\rec_postprocess.py", line 130, in decode
    char_list = [self.character[text_id] for text_id in text_index[batch_idx][selection]]
  File "E:\PaddleOCR-release-2.6\ppocr\postprocess\rec_postprocess.py", line 130, in <listcomp>
    char_list = [self.character[text_id] for text_id in text_index[batch_idx][selection]]
IndexError: list index out of range

### 🏃‍♂️ Environment (运行环境)

paddle 2.6.0
python 3.8
system:window 11
platsfrom: pycharm

### 🌰 Minimal Reproducible Example (最小可复现问题的Demo)

(paddle2) E:\PaddleOCR-release-2.6>python tools/infer/predict_system.py --image_dir="C:\Users\Administrator\Desktop\6.jpg" --det_model_dir="./inference_model/det/" --rec_model_dir="./inference_model/rec/"
[2024/11/25 12:21:22] ppocr INFO: In PP-OCRv3, rec_image_shape parameter defaults to '3, 48, 320', if you are using recognition model with PP-OCRv2 or an older version, please set --rec_image_shape='3,32,320
[2024/11/25 12:21:24] ppocr DEBUG: dt_boxes num : 11, elapsed : 1.905656337738037
Traceback (most recent call last):
  File "tools/infer/predict_system.py", line 272, in <module>
    main(args)
  File "tools/infer/predict_system.py", line 190, in main
    dt_boxes, rec_res, time_dict = text_sys(img)
  File "tools/infer/predict_system.py", line 105, in __call__
    rec_res, elapse = self.text_recognizer(img_crop_list)
  File "E:\PaddleOCR-release-2.6\tools\infer\predict_rec.py", line 623, in __call__
    rec_result = self.postprocess_op(preds, return_word_box=self.return_word_box, wh_ratio_list=wh_ratio_list, max_wh_ratio=max_wh_ratio)
  File "E:\PaddleOCR-release-2.6\ppocr\postprocess\rec_postprocess.py", line 170, in __call__
    text = self.decode(preds_idx, preds_prob, is_remove_duplicate=True, return_word_box=return_word_box)
  File "E:\PaddleOCR-release-2.6\ppocr\postprocess\rec_postprocess.py", line 130, in decode
    char_list = [self.character[text_id] for text_id in text_index[batch_idx][selection]]
  File "E:\PaddleOCR-release-2.6\ppocr\postprocess\rec_postprocess.py", line 130, in <listcomp>
    char_list = [self.character[text_id] for text_id in text_index[batch_idx][selection]]
IndexError: list index out of range

**Comment by GreatV** [comment]
试试新版的paddleocr

**Comment by Advance-Mr** [comment]
> 试试新版的paddleocr

换了张图片。成这样了：[2024/11/25 12:48:19] ppocr DEBUG: dt_boxes num : 0, elapsed : 1.5045967102050781
[2024/11/25 12:48:19] ppocr DEBUG: rec_res num  : 0, elapsed : 0.0
[2024/11/25 12:48:19] ppocr DEBUG: 0  Predict time of C:\Users\Administrator\Desktop\test.jpg: 1.506s
[2024/11/25 12:48:19] ppocr DEBUG: The visualized image saved in ./inference_results\test.jpg
[2024/11/25 12:48:19] ppocr INFO: The predict total time is 1.5125854015350342

**Comment by GreatV** [comment]
根据提供的信息和错误描述，您的问题出在识别模型（rec）推理阶段，具体是 `IndexError: list index out of range`。以下是对问题的详细分析和解决方法：

---

### **问题分析**

1. **错误日志关键点：**
   - 报错发生在 `rec_postprocess.py` 的 `decode` 函数中：
     ```python
     char_list = [self.character[text_id] for text_id in text_index[batch_idx][selection]]
     ```
     错误提示 `IndexError: list index out of range` 表示访问了超出 `self.character` 列表大小的索引值。
   - 这通常表明模型输出的预测索引 `text_index` 包含了非法值（超出字典范围）。

2. **可能的原因：**
   - **字典文件问题：**
     如果训练模型时使用的字典文件（`character_dict_path`）与推理时指定的字典文件不一致，可能会导致模型输出的索引值超出字典范围。
   - **识别模型版本不匹配：**
     如果您使用的是 PP-OCRv3 的识别模型，但未正确设置 `rec_image_shape` 参数，可能会导致模型输入形状错误，从而引起推理错误。
   - **模型文件损坏：**
     如果推理模型（`inference.pdmodel` 和 `inference.pdiparams`）在转换或下载过程中损坏，也可能导致预测结果异常。

---

### **解决方案**

#### 1. **确认字典文件一致性**
   - 检查训练模型时使用的字典文件路径，确保与推理时通过 `--rec_char_dict_path` 指定的路径一致。
   - 默认的中文字典文件路径通常是：
     ```bash
     ./ppocr/utils/ppocr_keys_v1.txt
     ```
   - 如果您在训练时自定义了字典文件，请确保推理时使用相同的文件。

#### 2. **验证 `rec_image_shape` 参数**
   - 根据错误日志，推理时使用了识别模型，但未明确设置 `rec_image_shape` 参数。
   - 对于 PP-OCRv3 模型，默认的 `rec_image_shape` 是 `[3, 48, 320]`。如果您使用的是旧版本（PP-OCRv2 或更早），需要将此参数改为 `[3, 32, 320]`。
   - 示例命令：
     ```bash
     python tools/infer/predict_system.py \
         --image_dir="C:\Users\Administrator\Desktop\6.jpg" \
         --det_model_dir="./inference_model/det/" \
         --rec_model_dir="./inference_model/rec/" \
         --rec_image_shape="3,48,320"
     ```

#### 3. **检查模型版本兼容性**
   - 确保检测模型（`det_model_dir`）和识别模型（`rec_model_dir`）属于同一版本的 PP-OCR。如果检测模型是 PP-OCRv3 的版本，而识别模型是 PP-OCRv2 的版本，可能会出现不兼容问题。
   - 如果确认版本不一致，请重新下载或训练匹配版本的模型。

#### 4. **重新导出推理模型**
   - 如果怀疑模型文件损坏，可以尝试重新导出推理模型。以下是导出命令示例：
     ```bash
     python3 tools/export_model.py \
         -c configs/rec/ch_PP-OCRv3_rec.yml \
         -o Global.pretrained_model=path/to/your/trained_model/best_accuracy \
         Global.save_inference_dir=./inference_model/rec/
     ```
   - 确保导出的模型文件目录下有以下三个文件：
     - `inference.pdmodel`
     - `inference.pdiparams`
     - `inference.pdiparams.info`

#### 5. **测试单独的识别模型**
   - 为了排查问题，可以单独测试识别模型，看看是否仍然出现错误：
     ```bash
     python tools/infer/predict_rec.py \
         --image_dir="C:\Users\Administrator\Desktop\6.jpg" \
         --rec_model_dir="./inference_model/rec/" \
         --rec_image_shape="3,48,320"
     ```
   - 如果此命令执行成功，说明问题可能出在检测和识别模型的联动部分；否则问题出在识别模型本身。

#### 6. **检查 PaddleOCR 版本**
   - 确保代码和模型兼容。建议使用与模型版本一致的 PaddleOCR 代码库（例如，PP-OCRv3 建议使用 `release/2.6` 分支）。更新代码：
     ```bash
     git checkout release/2.6
     ```

#### 7. **调试日志**
   - 设置日志级别为 DEBUG，以便获取更多详细信息：
     ```bash
     python tools/infer/predict_system.py \
         --image_dir="C:\Users\Administrator\Desktop\6.jpg" \
         --det_model_dir="./inference_model/det/" \
         --rec_model_dir="./inference_model/rec/" \
         --rec_image_shape="3,48,320" \
         --show_log=True
     ```

---

### **总结**
根据错误日志，问题的根本原因可能是：
1. 字典文件不一致。
2. `rec_image_shape` 参数未正确设置，导致模型输入形状错误。
3. 模型版本不匹配或文件损坏。

建议按照上述步骤逐一排查，确保模型和配置参数的一致性。如果问题仍未解决，可以在 GitHub 提交 Issue 并附上详细日志和配置文件以获得进一步支持。

---
*Response generated by 🤖 feifei-bot | chatgpt-4o-latest*

**Comment by GreatV** [comment]
@Advance-Mr 试试paddleocr 2.8.1 看看还有没有这个问题

**Comment by Advance-Mr** [comment]
> @Advance-Mr 试试paddleocr 2.8.1 看看还有没有这个问题

好的。根据你的回复，我的det是ch_det_res18_db_v2.0.yml，rec是chinese_cht_PP-OCRv3_rec.yml。
我先试281version. 不行我在尝试将det和rec换做同一版本。待会给你答复

**Comment by Advance-Mr** [comment]
rec的部分yml如下：
 checkpoints:
  save_inference_dir:
  use_visualdl: false
  infer_img: doc/imgs_words/chinese_traditional/chinese_cht_2.png
  character_dict_path:  E:/PaddleOCR-release-2.6/ppocr/utils/dict/chinese_cht_dict.txt
  max_text_length: &max_text_length 100
  infer_mode: false
  use_space_char: true
  distributed: true
  save_res_path: ./output/rec/chinese_cht.txt

**Comment by Advance-Mr** [comment]
我使用ch_PP-OCRv3_det_cml.yml
cal_metric_during_train: false
  pretrained_model: ./pretrain_models/Multilingual_PP-OCRv3_det_distill_train/best_accuracy
  checkpoints: ./output/ch_PP-OCR_v3_det/latest
  save_inference_dir: null
  use_visualdl: false
  infer_img: doc/imgs_en/img_10.jpg
  save_res_path: ./output/det_db/predicts_db.txt
  distributed: true

Architecture:
  name: DistillationModel
  algorithm: Distillation
  model_type: det
  Models:
    Student:
      pretrained:
      model_type: det
      algorithm: DB
      Transform: null
      Backbone:
        name: MobileNetV3
        scale: 0.5
        model_name: large
        disable_se: true
      Neck:
        name: RSEFPN
        out_channels: 96
        shortcut: True
      Head:
        name: DBHead
        k: 50
    Student2:
      pretrained:
      model_type: det
      algorithm: DB
      Transform: null
      Backbone:
        name: MobileNetV3
        scale: 0.5
        model_name: large
        disable_se: true
      Neck:
        name: RSEFPN
        out_channels: 96
        shortcut: True
      Head:
        name: DBHead
        k: 50
    Teacher:
      freeze_params: true
      return_all_feats: false
      model_type: det
      algorithm: DB
      Backbone:
        name: ResNet_vd
        in_channels: 3
        layers: 50
      Neck:
        name: LKPAN
        out_channels: 256
      Head:
        name: DBHead
        kernel_list: [7,2,2]
        k: 50

Loss:
  name: CombinedLoss
  loss_config_list:
  - DistillationDilaDBLoss:
      weight: 1.0
      model_name_pairs:
      - ["Student", "Teacher"]
      - ["Student2", "Teacher"]
      key: maps
      balance_loss: true
      main_loss_type: DiceLoss
      alpha: 5
      beta: 10
      ohem_ratio: 3
  - DistillationDMLLoss:
      model_name_pairs:
      - ["Student", "Student2"]
      maps_name: "thrink_maps"
      weight: 1.0
      model_name_pairs: ["Student", "Student2"]
      key: maps
  - DistillationDBLoss:
      weight: 1.0
      model_name_list: ["Student", "Student2"]
      balance_loss: true
      main_loss_type: DiceLoss
      alpha: 5
      beta: 10
      ohem_ratio: 3

Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Cosine
    learning_rate: 0.001
    warmup_epoch: 2
  regularizer:
    name: L2
    factor: 5.0e-05

PostProcess:
  name: DistillationDBPostProcess
  model_name: ["Student"]
  key: head_out
  thresh: 0.3
  box_thresh: 0.6
  max_candidates: 1000
  unclip_ratio: 1.5

Metric:
  name: DistillationMetric
  base_metric_name: DetMetric
  main_indicator: hmean
  key: "Student"

Train:
  dataset:
    name: SimpleDataSet
    data_dir: ./train_data/det/
    label_file_list:
      - ./train_data/det/train.txt
    ratio_list: [1.0]
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - DetLabelEncode: null
    - CopyPaste:
    - IaaAugment:
        augmenter_args:
        - type: Fliplr
          args:
            p: 0.5
        - type: Affine
          args:
            rotate:
            - -10
            - 10
        - type: Resize
          args:
            size:
            - 0.5
            - 3
    - EastRandomCropData:
        size:
        - 960
        - 960
        max_tries: 50
        keep_ratio: true
    - MakeBorderMap:
        shrink_ratio: 0.4
        thresh_min: 0.3
        thresh_max: 0.7
    - MakeShrinkMap:
        shrink_ratio: 0.4
        min_text_size: 8
    - NormalizeImage:
        scale: 1./255.
        mean:
        - 0.485
        - 0.456
        - 0.406
        std:
        - 0.229
        - 0.224
        - 0.225
        order: hwc
    - ToCHWImage: null
    - KeepKeys:
        keep_keys:
        - image
        - threshold_map
        - threshold_mask
        - shrink_map
        - shrink_mask
  loader:
    shuffle: true
    drop_last: false
    batch_size_per_card: 1
    num_workers: 1

Eval:
  dataset:
    name: SimpleDataSet
    data_dir: ./train_data/det/
    label_file_list:
      - ./train_data/det/val.txt
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - DetLabelEncode: # Class handling label
      - DetResizeForTest:
      - NormalizeImage:
          scale: 1./255.
          mean: [0.485, 0.456, 0.406]
          std: [0.229, 0.224, 0.225]
          order: 'hwc'
      - ToCHWImage:
      - KeepKeys:
          keep_keys: ['image', 'shape', 'polys', 'ignore_tags']
  loader:
    shuffle: False
    drop_last: False
    batch_size_per_card: 1 # must be 1
    num_workers: 1 训练出的det可以正确框出文字。但.txt文件内容是
C:\Users\Administrator\Desktop\3.jpg	[]  没有数据？

**Comment by GreatV** [comment]
是不是字典没有设置，`rec_char_dict_path` 推理的时候要设置对应的字典。

**Comment by Advance-Mr** [comment]
> 是不是字典没有设置，`rec_char_dict_path` 推理的时候要设置对应的字典
嗯，是这样的，我这个是det模型，识别完后，用一张图片测试，会框出文本，但txt中没有坐标值。你说是rec模型，在训练过程中我加了这个字典，测试是通过的。

**Comment by GreatV** [comment]
就是 `character_dict_path`，我的意思是rec推理的时候需要配置成你训练rec 模型时用的字典。

**Comment by GreatV** [comment]
新的问题重新开个issue问吧
